R/cell_labeling.R

```diff
@@ -0,0 +1,223 @@
+#' @title Calculate normalized expression of a gene set
+#' 
+#' @description Calculate normalized expression for a set of genes in each cell
+#'   from a \code{SingleCellExperiment}, using random sets of similarly
+#'   expressed genes as background to account for cell quality and
+#'   sequencing depth.
+#'
+#' @author Michael Stadler
+#'
+#' @param sce \code{SingleCellExperiment} object.
+#' @param genes \code{character} vector with the genes in the set. Must be a
+#'   subset of \code{rownames(sce)}.
+#' @param expr_values Integer scalar or string indicating which assay of
+#'   \code{sce} contains the expression values.
+#' @param R Integer scalar giving the number of random gene sets to sample
+#'   for normalization.
+#' @param BPPARAM An optional \code{\link[BiocParallel]{BiocParallelParam}}
+#'   instance determining the parallel back-end to be used during evaluation.
+#' 
+#' @return A \code{numeric} vector with normalized gene set scores for each
+#'   cell in \code{sce}.
+#' 
+#' @importFrom BiocParallel SerialParam bplapply
+#' @importFrom SummarizedExperiment assay assayNames
+#' @importFrom matrixStats rowMeans2 rowSds
+#' 
+#' @export
+normGenesetExpression <- function(sce,
+                                  genes,
+                                  expr_values = "logcounts",
+                                  R = 200,
+                                  BPPARAM = SerialParam()) {
+    # pre-flight checks
+    stopifnot(exprs = {
+        is(sce, "SingleCellExperiment")
+        is.character(genes)
+        length(genes) > 1L
+        all(genes %in% rownames(sce))
+        is.numeric(R)
+        length(R) == 1L
+        R > 0
+        inherits(BPPARAM, "BiocParallelParam")
+    })
+    if (is.numeric(expr_values) && length(expr_values) == 1L)
+        expr_values <- assayNames(sce)[expr_values]
```

Comment:
This is a bit dangerous I think, since assays can be unnamed, or there can be multiple assays with the same name. So if a numeric index is allowed and specified (which would unambiguously identify the assay), it's probably best to just keep it that way.

Reply:
Ok, I had assumed that `assayNames` are unique and non-`NULL`, but in that case I'll change the code accordingly.